man/ppdmerge.1
.\"
.\" ppdmerge man page for CUPS.
.\"
.\" Copyright © 2020-2025 by OpenPrinting.
.\" Copyright © 2007-2019 by Apple Inc.
.\" Copyright © 1997-2007 by Easy Software Products.
.\"
.\" Licensed under Apache License v2.0.  See the file "LICENSE" for more
.\" information.
.\"
.TH ppdmerge 1 "CUPS" "2021-02-28" "OpenPrinting"
.SH NAME
ppdmerge \- merge ppd files (deprecated)
.SH SYNOPSIS
.B ppdmerge
[
.B \-o
.I output-ppd-file
]
.I ppd-file
.I ppd-file
[ ...
.I ppd-file
]
.SH DESCRIPTION
\fBppdmerge\fR merges two or more PPD files into a single, multi-language
PPD file.
\fBThis program is deprecated and will be removed in a future release of CUPS.\fR
.SH OPTIONS
\fBppdmerge\fR supports the following options:
.TP
\fB\-o \fIoutput-ppd-file\fR
Specifies the PPD file to create.
If not specified, the merged PPD file is written to the standard output.
If the output file already exists, it is silently overwritten.
.SH NOTES
PPD files are deprecated and will no longer be supported in a future feature release of CUPS.
Printers that do not support IPP can be supported using applications such as
.BR ippeveprinter (1).
.LP
\fBppdmerge\fR does not check whether the merged PPD files are for the same device.
Merging of different device PPDs will yield unpredictable results.
.SH SEE ALSO
.BR ppdc (1),
.BR ppdhtml (1),
.BR ppdi (1),
.BR ppdpo (1),
.BR ppdcfile (5),
CUPS Online Help (http://localhost:631/help)
.SH COPYRIGHT
Copyright \[co] 2020-2025 by OpenPrinting.


CHANGES.md
CHANGES - OpenPrinting CUPS
===========================


Changes in CUPS v2.4.12 (YYYY-MM-DD)
------------------------------------

- GnuTLS follows system crypto policies now (Issue #1105)
- Added `NoSystem` SSLOptions value (Issue #1130)
- Fixed a compressed file error handling bug (Issue #1070)
- Fixed a bug in the make-and-model whitespace trimming code (Issue #1096)
- Fixed a removal of IPP Everywhere permanent queue if installation failed (Issue #1102)
- Fixed `ServerToken None` in scheduler (Issue #1111)
- Fixed the default User-Agent string.
- Fixed a recursion issue in `ippReadIO`.


Changes in CUPS v2.4.11 (2024-09-30)
------------------------------------

- Updated the maximum file descriptor limit for `cupsd` to 64k-1 (Issue #989)
- Fixed `lpoptions -d` with a discovered but not added printer (Issue #833)
- Fixed incorrect error message for HTTP/IPP errors (Issue #893)
- Fixed JobPrivateAccess and SubscriptionPrivateAccess support for "all"
  (Issue #990)
- Fixed issues with cupsGetDestMediaByXxx (Issue #993)
- Fixed adding and modifying of printers via the web interface (Issue #998)
- Fixed HTTP PeerCred authentication for domain users (Issue #1001)
- Fixed checkbox support (Issue #1008)
- Fixed printer state notifications (Issue #1013)
- Fixed IPP Everywhere printer setup (Issue #1033)


Changes in CUPS v2.4.10 (2024-06-18)
------------------------------------

- Fixed error handling when reading a mixed `1setOf` attribute.
- Fixed scheduler start if there is only domain socket to listen on (Issue #985)


Changes in CUPS v2.4.9 (2024-06-11)
-----------------------------------

- Fixed domain socket handling (CVE-2024-35235)
- Fixed creating of `cupsUrfSupported` PPD keyword (Issue #952)
- Fixed searching for destinations in web ui (Issue #954)
- Fixed TLS negotiation using OpenSSL with servers that require the TLS SNI
  extension.
- Really raised `cups_enum_dests()` timeout for listing available IPP printers
  (Issue #751)...
- Fixed `Host` header regression (Issue #967)
- Fixed DNS-SD lookups of local services with Avahi (Issue #970)
- Fixed listing jobs in destinations in web ui. (Apple issue #6204)
- Fixed showing search query in web ui help page. (Issue #977)


Changes in CUPS v2.4.8 (2024-04-26)
-----------------------------------

- Added warning if the device has to be asked for 'all,media-col-database'
  separately (Issue #829)
- Added new value for 'lpstat' option '-W' - successfull - for getting
  successfully printed jobs (Issue #830)
- Added support for PAM modules password-auth and system-auth (Issue #892)
- Updated IPP Everywhere printer creation error reporting (Issue #347)
- Updated and documented the MIME typing buffering limit (Issue #925)
- Now report an error for temporary printer defaults with lpadmin (Issue #237)
- Fixed mapping of PPD InputSlot, MediaType, and OutputBin values (Issue #238)
- Fixed "document-unprintable-error" handling (Issue #391)
- Fixed the web interface not showing an error for a non-existent printer
  (Issue #423)
- Fixed printing of jobs with job name longer than 255 chars on older printers
  (Issue #644)
- Really backported fix for Issue #742
- Fixed `cupsCopyDestInfo` device connection detection (Issue #586)
- Fixed "Upgrade" header handling when there is no TLS support (Issue #775)
- Fixed memory leak when unloading a job (Issue #813)
- Fixed memory leak when creating color profiles (Issue #815)
- Fixed a punch finishing bug in the IPP Everywhere support (Issue #821)
- Fixed crash in `scan_ps()` if incoming argument is NULL (Issue #831)
- Fixed setting job state reasons for successful jobs (Issue #832)
- Fixed infinite loop in IPP backend if hostname is IP address with Kerberos
  (Issue #838)
- Added additional check on socket if `revents` from `poll()` returns POLLHUP
  together with POLLIN or POLLOUT in `httpAddrConnect2()` (Issue #839)
- Fixed crash in `ppdEmitString()` if `size` is NULL (Issue #850)
- Fixed reporting `media-source-supported` when sharing printer which has
  numbers as strings instead of keywords as `InputSlot` values (Issue #859)
- Fixed IPP backend to support the "print-scaling" option with IPP printers
  (Issue #862)
- Fixed potential race condition for the creation of temporary queues
  (Issue #871)
- Fixed `httpGets` timeout handling (Issue #879)
- Fixed checking for required attributes during PPD generation (Issue #890)
- Fixed encoding of IPv6 addresses in HTTP requests (Issue #903)
- Fixed sending response headers to client (Issue #927)
- Fixed CGI program initialization and validation of form checkbox and text
  fields.


Changes in CUPS v2.4.7 (2023-09-20)
-----------------------------------

- CVE-2023-4504 - Fixed Heap-based buffer overflow when reading Postscript
  in PPD files
- Added OpenSSL support for cupsHashData (Issue #762)
- Fixed delays in lpd backend (Issue #741)
- Fixed extensive logging in scheduler (Issue #604)
- Fixed hanging of `lpstat` on IBM AIX (Issue #773)
- Fixed hanging of `lpstat` on Solaris (Issue #156)
- Fixed printing to stderr if we can't open cups-files.conf (Issue #777)
- Fixed purging job files via `cancel -x` (Issue #742)
- Fixed RFC 1179 port reserving behavior in LPD backend (Issue #743)
- Fixed a bug in the PPD command interpretation code (Issue #768)
- Fixed Oki 407 freeze when printing larger jobs (Issue #877)


Changes in CUPS v2.4.6 (2023-06-22)
-----------------------------------

- CVE-2023-34241: Fixed use-after-free when logging warnings in case of failures
  in `cupsdAcceptClient()`.
- Fixed linking error on old MacOS (Issue #715)
- Fixed printing multiple files on specific printers (Issue #643)


Changes in CUPS v2.4.5 (2023-06-13)
-----------------------------------

- Fixed corruption of locally saved certificates (Issue #724)


Changes in CUPS v2.4.4 (2023-06-06)
-----------------------------------

- Fixed segfault in `cupsGetNamedDest()` when trying to get default printer, but
  the default printer is not set (Issue #719)



ppdc/ppdmerge.cxx
//
// PPD file merge utility for the CUPS PPD Compiler.
//
// Copyright © 2020-2025 by OpenPrinting.
// Copyright © 2007-2018 by Apple Inc.
// Copyright © 2002-2007 by Easy Software Products.
//
// Licensed under Apache License v2.0.  See the file "LICENSE" for more
// information.
//

//
// Include necessary headers...
//

#include <cups/cups-private.h>
#include <cups/ppd-private.h>
#include <cups/array.h>


//
// Local functions...
//

static const char	*ppd_locale(ppd_file_t *ppd);
static void		usage(void) _CUPS_NORETURN;


//
// 'main()' - Main entry for the PPD merge utility.
//

int					// O - Exit status
main(int  argc,				// I - Number of command-line arguments
     char *argv[])			// I - Command-line arguments
{
  int		i;			// Looping var
  char		*opt;			// Current option
  ppd_file_t	*ppd;			// PPD file
  cups_array_t	*ppds;			// Array of PPD files
  const char	*inname,		// First input filename
		*outname;		// Output filename (if any)
  char		bckname[1024];		// Backup filename
  cups_file_t	*infile,		// Input file
		*outfile;		// Output file
  cups_array_t	*languages;		// Languages in file
  const char	*locale;		// Current locale
  char		line[1024];		// Line from file


  _cupsSetLocale(argv);

  // Scan the command-line...
  inname    = NULL;
  outname   = NULL;
  outfile   = NULL;
  languages = NULL;
  ppds      = cupsArrayNew(NULL, NULL);

  for (i = 1; i < argc; i ++)
    if (argv[i][0] == '-')
    {
      for (opt = argv[i] + 1; *opt; opt ++)
        switch (*opt)
	{
	  case 'o' :			// Output file
              if (outname)
	        usage();

	      i ++;
	      if (i >= argc)
        	usage();

	      outname = argv[i];
	      break;

	  default :			// Unknown
	      usage();
        }
    }
    else
    {
      // Open and load the PPD file...
      if ((infile = cupsFileOpen(argv[i], "r")) == NULL)
      {
        _cupsLangPrintf(stderr, _("%s: Unable to open %s: %s"), "ppdmerge",
	                argv[i], strerror(errno));
	return (1);
      }

      // Open the PPD file...
      if ((ppd = ppdOpen2(infile)) == NULL)
      {
        ppd_status_t	status;		// PPD open status
	int		curline,	// Current line
			linenum;	// Line number


        status = ppdLastError(&linenum);

	_cupsLangPrintf(stderr,
	                _("%s: Unable to open PPD file: %s on line %d."),
	                "ppdmerge", ppdErrorString(status), linenum);
        cupsFileRewind(infile);

        line[0] = '\0';
	curline = 0;

        while (cupsFileGets(infile, line, sizeof(line)))
	{
	  curline ++;
	  if (curline >= linenum)
	    break;
	}

	_cupsLangPrintf(stderr, "%d: %s", linenum, line);

        cupsFileClose(infile);
	return (1);
      }

      // Figure out the locale...
      if ((locale = ppd_locale(ppd)) == NULL)
      {
        _cupsLangPrintf(stderr,
	                _("ppdmerge: Bad LanguageVersion \"%s\" in %s."),
			ppd->lang_version, argv[i]);
        cupsFileClose(infile);
	ppdClose(ppd);
	return (1);
      }

      if (!strcmp(locale, "en") && !inname && !outfile)
      {
        // Set the English PPD's filename...
	inname    = argv[i];
	languages = _ppdGetLanguages(ppd);

        if (outname && !strcmp(inname, outname))
	{
	  // Rename input filename so that we don't overwrite it...
	  snprintf(bckname, sizeof(bckname), "%s.bck", inname);

	  if (rename(inname, bckname))
	  {
	    _cupsLangPrintf(stderr,
	                    _("ppdmerge: Unable to backup %s to %s - %s"),
			    inname, bckname, strerror(errno));
	    return (1);
	  }

	  inname = bckname;
	}
      }
      else if (strcmp(locale, "en"))
      {
	// Save this PPD for later processing...
        cupsArrayAdd(ppds, ppd);
      }
      else
      {
        // Don't need this PPD...
	_cupsLangPrintf(stderr, _("ppdmerge: Ignoring PPD file %s."),
	                argv[i]);
        ppdClose(ppd);
      }

      // Close and move on...
      cupsFileClose(infile);
    }

  // If no PPDs have been loaded, display the program usage message.
  if (!inname)
    usage();

  // Loop through the PPD files we loaded to generate a new language list...
  if (!languages)
    languages = cupsArrayNew((cups_array_func_t)strcmp, NULL);

  for (ppd = (ppd_file_t *)cupsArrayFirst(ppds);
       ppd;
       ppd = (ppd_file_t *)cupsArrayNext(ppds))
  {
    locale = ppd_locale(ppd);

    if (cupsArrayFind(languages, (void *)locale))
    {
      // Already have this language, remove the PPD from the list.
      ppdClose(ppd);
      cupsArrayRemove(ppds, ppd);
    }
    else
      cupsArrayAdd(languages, (void *)locale);
  }

  // Copy the English PPD starting with a cupsLanguages line...
  infile = cupsFileOpen(inname, "r");

  if (outname)
  {
    const char *ext = strrchr(outname, '.');
    if (ext && !strcmp(ext, ".gz"))
      outfile = cupsFileOpen(outname, "w9");
    else
      outfile = cupsFileOpen(outname, "w");
  }
  else
    outfile = cupsFileStdout();

  cupsFileGets(infile, line, sizeof(line));
  cupsFilePrintf(outfile, "%s\n", line);
  if ((locale = (char *)cupsArrayFirst(languages)) != NULL)
  {
    cupsFilePrintf(outfile, "*cupsLanguages: \"%s", locale);
    while ((locale = (char *)cupsArrayNext(languages)) != NULL)
      cupsFilePrintf(outfile, " %s", locale);
    cupsFilePuts(outfile, "\"\n");
  }

  while (cupsFileGets(infile, line, sizeof(line)))
  {
    if (strncmp(line, "*cupsLanguages:", 15))
      cupsFilePrintf(outfile, "%s\n", line);
  }

  // Loop through the other PPD files we loaded to provide the translations...
  for (ppd = (ppd_file_t *)cupsArrayFirst(ppds);
       ppd;
       ppd = (ppd_file_t *)cupsArrayNext(ppds))
  {
    // Output all of the UI text for this language...
    int			j, k, l;	// Looping vars
    ppd_group_t		*g;		// Option group
    ppd_option_t	*o;		// Option
    ppd_choice_t	*c;		// Choice
    ppd_coption_t	*co;		// Custom option
    ppd_cparam_t	*cp;		// Custom parameter
    ppd_attr_t		*attr;		// PPD attribute

    locale = ppd_locale(ppd);

    cupsFilePrintf(outfile, "*%% %s localization\n", ppd->lang_version);
    cupsFilePrintf(outfile, "*%s.Translation ModelName/%s: \"\"\n", locale,
		   ppd->modelname);

    for (j = ppd->num_groups, g = ppd->groups; j > 0; j --, g ++)
    {
      cupsFilePrintf(outfile, "*%s.Translation %s/%s: \"\"\n", locale,
		     g->name, g->text);

      for (k = g->num_options, o = g->options; k > 0; k --, o ++)
      {
	cupsFilePrintf(outfile, "*%s.Translation %s/%s: \"\"\n", locale,
		       o->keyword, o->text);

	for (l = o->num_choices, c = o->choices; l > 0; l --, c ++)
	  cupsFilePrintf(outfile, "*%s.%s %s/%s: \"\"\n", locale,
			 o->keyword, c->choice, c->text);

	if ((co = ppdFindCustomOption(ppd, o->keyword)) != NULL)
	{
	  snprintf(line, sizeof(line), "Custom%s", o->keyword);
	  attr = ppdFindAttr(ppd, line, "True");
	  cupsFilePrintf(outfile, "*%s.Custom%s True/%s: \"\"\n", locale,
			 o->keyword, attr->text);
	  for (cp = ppdFirstCustomParam(co); cp; cp = ppdNextCustomParam(co))
	    cupsFilePrintf(outfile, "*%s.ParamCustom%s %s/%s: \"\"\n", locale,
			   o->keyword, cp->name, cp->text);
	}
      }
    }

    ppdClose(ppd);
  }

  cupsArrayDelete(ppds);

  cupsFileClose(outfile);

  // Return with no errors.
  return (0);
}


//
// 'ppd_locale()' - Return the locale associated with a PPD file.
//

static const char *			// O - Locale string
ppd_locale(ppd_file_t *ppd)		// I - PPD file
{
  int		i;			// Looping var
  size_t	vlen;			// Length of LanguageVersion string
  static char	locale[255];		// Locale string
  static struct				// LanguageVersion translation table
  {
    const char	*version,		// LanguageVersion string */
		*language;		// Language code */
  }		languages[] =
  {
    { "chinese",		"zh" },
    { "czech",			"cs" },
    { "danish",			"da" },
    { "dutch",			"nl" },
    { "english",		"en" },
    { "finnish",		"fi" },
    { "french",			"fr" },
    { "german",			"de" },
    { "greek",			"el" },
    { "hungarian",		"hu" },
    { "italian",		"it" },
    { "japanese",		"ja" },
    { "korean",			"ko" },
    { "norwegian",		"no" },
    { "polish",			"pl" },
    { "portuguese",		"pt" },
    { "russian",		"ru" },
    { "simplified chinese",	"zh_CN" },
    { "slovak",			"sk" },
    { "spanish",		"es" },
    { "swedish",		"sv" },
    { "traditional chinese",	"zh_TW" },
    { "turkish",		"tr" }
  };


  for (i = 0; i < (int)(sizeof(languages) / sizeof(languages[0])); i ++)
  {
    vlen = strlen(languages[i].version);

    if (!_cups_strncasecmp(ppd->lang_version, languages[i].version, vlen))
    {
      if (ppd->lang_version[vlen] == '-' ||
          ppd->lang_version[vlen] == '_')
        snprintf(locale, sizeof(locale), "%s_%s", languages[i].language,
	         ppd->lang_version + vlen + 1);
      else
        strlcpy(locale, languages[i].language, sizeof(locale));

      return (locale);
    }
  }

  return (NULL);
}

//
// 'usage()' - Show usage and exit.
//

static void
usage(void)
{
  _cupsLangPuts(stdout, _("Usage: ppdmerge [options] filename.ppd [ ... "
                          "filenameN.ppd ]"));
  _cupsLangPuts(stdout, _("Options:"));
  _cupsLangPuts(stdout, _("  -o filename.ppd[.gz]    Set output file "
                          "(otherwise stdout)."));

  exit(1);
}
